controllers/topology/cluster_controller_test.go

```diff
@@ -283,51 +286,143 @@ func TestClusterReconciler_reconcileUpdatesOnClusterClass(t *testing.T) {
+	g.Expect(err).ToNot(HaveOccurred())
+
+	// Defer a cleanup function that deletes each of the objects created during setupTestEnvForIntegrationTests.
+	defer func() {
+		g.Expect(cleanup()).To(Succeed())
+	}()
+
+	actualCluster := &clusterv1.Cluster{}
+	// First ensure that the initial cluster and other objects are created and populated as expected.
+	g.Eventually(func(g Gomega) error {
+		// Get the cluster object now including the updated replica number for the Machine deployment.
+		if err := env.Get(ctx, client.ObjectKey{Name: clusterName1, Namespace: ns.Name}, actualCluster); err != nil {
+			return err
+		}
+
+		// Check if Cluster has relevant Infrastructure and ControlPlane and labels and annotations.
+		g.Expect(assertClusterReconcile(actualCluster)).Should(Succeed())
+
+		// Check if InfrastructureCluster has been created and has the correct labels and annotations.
+		g.Expect(assertInfrastructureClusterReconcile(actualCluster)).Should(Succeed())
+
+		// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
+		g.Expect(assertControlPlaneReconcile(actualCluster)).Should(Succeed())
+
+		// Check if MachineDeployments are created and have the correct version, replicas, labels annotations and templates.
+		g.Expect(assertMachineDeploymentsReconcile(actualCluster)).Should(Succeed())
+		return nil
+	}, timeout).Should(Succeed())
+
+	// Change the ClusterClass pointed to in the Cluster's Topology. This is a ClusterClass rebase operation.
+	clusterWithRebase := actualCluster.DeepCopy()
+	clusterWithRebase.Spec.Topology.Class = clusterClassName2
+	g.Expect(env.Update(ctx, clusterWithRebase)).Should(Succeed())
+
+	// Check to ensure all objects are correctly reconciled with the new ClusterClass.
+	g.Eventually(func(g Gomega) error {
+		// Get the cluster object.
+		updatedCluster := &clusterv1.Cluster{}
+		if err := env.Get(ctx, client.ObjectKey{Name: clusterName1, Namespace: ns.Name}, updatedCluster); err != nil {
+			return err
+		}
+		// Check to ensure the spec.topology.class has been successfully updated in the API server and cache.
+		g.Expect(updatedCluster.Spec.Topology.Class).To(Equal(clusterClassName2))
+
+		// Check if Cluster has relevant Infrastructure and ControlPlane and labels and annotations.
+		g.Expect(assertClusterReconcile(updatedCluster)).Should(Succeed())
+
+		// Check if InfrastructureCluster has been created and has the correct labels and annotations.
+		g.Expect(assertInfrastructureClusterReconcile(updatedCluster)).Should(Succeed())
+
+		// Check if ControlPlane has been created and has the correct version, replicas, labels and annotations.
+		g.Expect(assertControlPlaneReconcile(updatedCluster)).Should(Succeed())
+
+		// Check if MachineDeployments are created and have the correct version, replicas, labels annotations and templates.
+		g.Expect(assertMachineDeploymentsReconcile(updatedCluster)).Should(Succeed())
+		return nil
```

Comment:
might be something for a follow up iteration but I think we should add something to check that rebase properly worked.
e.g. 
- InfrastructureCluster should have the same name as before, but with the changes
- ControlPlane should have rotated the machineTemplate

Reply:
I think I'll create an issue for this, there's probably useful work to be done to make these tests pick up properly on spec changes in the templates.

Reply:
https://github.com/kubernetes-sigs/cluster-api/issues/5699